### PR TITLE
Sharing Datasets Across Process Boundaries

### PR DESCRIPTION
**Pull request by rusty1s**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/10806 by adding this information to the documentation.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: docs/source/advanced/training_tricks.rst [hunk 1]
```diff
@@ -153,3 +153,46 @@ Advanced GPU Optimizations
...
+        def setup(self, stage: Optional[str] = None):
+            self.mnist = MNIST(self.data_dir)
+
+        def train_loader(self):
+            return DataLoader(self.mnist, batch_size=128)
+
+However, for in-memory datasets, that means that each process will hold a (redundant) replica of the dataset in memory, which may be impractical when using many processes and utilizing datasets that nearly fit into CPU memory, as the memory consumption will scale up linearly with the number of processes.
+For example, when training Graph Neural Networks, a common strategy is to load the entire graph into CPU memory for fast access to the entire graph structure and its features, and perform neighbor sampling to obtain mini-batches that fit onto the GPU.
+
+A simple way to disallow redundant dataset replicas is to rely on :obj:`torch.multiprocessing` to share the `data automatically between spawned processes via shared memory <https://pytorch.org/docs/stable/multiprocessing.html>`_.
```
Comment (justusschock) [inline docs/source/advanced/training_tricks.rst:178]: This can be a bit troublesome depending on how many individual tensors you have. Depending on the type of communication and the number of tensors this can also lead to errors with too many open file handles. Just wanted to mention this so that we don't advertise this as the best way at all circumstances.

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source/advanced/training_tricks.rst (hunk 1 above)
Comment (rusty1s) [reply docs/source/advanced/training_tricks.rst:178]: Sounds great. Thanks for the feedback. Let me add a note and modify the example according (i.e. sharing a single tensor instead of many small individual ones).

Comment (rusty1s) [reply docs/source/advanced/training_tricks.rst]: Updated the doc according to your suggestions. Hope this resolve your concerns :)

**Review by justusschock — APPROVED** [review]

File: docs/source/advanced/training_tricks.rst [hunk 2]
```diff
@@ -153,3 +153,51 @@ Advanced GPU Optimizations
...
+            self.mnist = MNIST(self.data_dir)
+
+        def train_loader(self):
+            return DataLoader(self.mnist, batch_size=128)
+
+However, for in-memory datasets, that means that each process will hold a (redundant) replica of the dataset in memory, which may be impractical when using many processes while utilizing datasets that nearly fit into CPU memory, as the memory consumption will scale up linearly with the number of processes.
+For example, when training Graph Neural Networks, a common strategy is to load the entire graph into CPU memory for fast access to the entire graph structure and its features, and to then perform neighbor sampling to obtain mini-batches that fit onto the GPU.
+
+A simple way to prevent redundant dataset replicas is to rely on :obj:`torch.multiprocessing` to share the `data automatically between spawned processes via shared memory <https://pytorch.org/docs/stable/notes/multiprocessing.html>`_.
+For this, all data pre-processing should be done on the main process inside :meth:`DataModule.__init__`.
```
Comment (awaelchli) [inline docs/source/advanced/training_tricks.rst:179]: data pre-processing can still be done inside the worker processes, right? So not all of it has to be done in the main process. Should we say "pre-loading" instead of "pre-processing"?

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@rusty1s one of our test workflows doesn't run. I think it is because you submitted a branch from an outdated fork. Could you try to rebase your branch on top of the latest master?

**Comment by rusty1s** [comment]
Yeah, it was indeed pretty out-dated. Thanks for spotting. Let's see if CI runs through now :)

**Comment by awaelchli** [comment]
Awesome, thanks for this addition @rusty1s
